staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go

```diff
@@ -0,0 +1,274 @@
+	assert.False(t, portExists)
+}
+
+func verifyV2Object(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, obj *unstructured.Unstructured) {
+	assert.Equal(t, crd.Spec.Group+"/v2", obj.GetAPIVersion())
+	_, hostPortExists := obj.Object["hostPort"]
+	assert.False(t, hostPortExists)
+	host, hostExists := obj.Object["host"]
+	assert.True(t, hostExists)
+	assert.Equal(t, "localhost", host)
+	port, portExists := obj.Object["port"]
+	assert.True(t, portExists)
+	assert.Equal(t, "8080", port)
+}
+
+func testCustomResourceConversionWebhook(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, customResourceClients map[string]dynamic.ResourceInterface) {
+	name := "cr-instance-1"
+	t.Log("Creating a v1 custom resource")
+	crInstance := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       crd.Spec.Names.Kind,
+			"apiVersion": crd.Spec.Group + "/v1",
+			"metadata": map[string]interface{}{
+				"name": name,
+			},
+			"hostPort": "localhost:8080",
+		},
+	}
+	_, err := customResourceClients["v1"].Create(crInstance, metav1.CreateOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	t.Log("v2 custom resource should be converted")
+	v2cr, err := customResourceClients["v2"].Get(name, metav1.GetOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	if !assert.NotNil(t, v2cr) {
+		t.FailNow()
+	}
+	verifyV2Object(t, crd, v2cr)
+}
+
+func testCRListConversion(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, apiextensionsClient apiextensionsclientset.Interface, customResourceClients map[string]dynamic.ResourceInterface) {
+	name1 := "cr-instance-1"
+	name2 := "cr-instance-2"
+	t.Log("Creating a v1 custom resource")
+
+	// cr-instance-2 is stored as v1. Now change storage version to v2.
+	var updateCRD *apiextensionsv1beta1.CustomResourceDefinition
+	var err error
+	for i := 0; i < 5; i++ {
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to get CustomResourceDefinition %q: %v", crd.Name, err)
+		}
+
+		// swap storage version
+		updateCRD.Spec.Versions[0].Storage = false
+		updateCRD.Spec.Versions[1].Storage = true
+
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Update(updateCRD)
+		if err == nil {
+			break
+		}
+		if !apierrors.IsConflict(err) {
+			break
+		}
+	}
+	if err != nil {
+		t.Fatal(err)
+	}
+	crd = updateCRD
+
+	t.Log("Create a v2 custom resource")
+	v1cr3 := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       crd.Spec.Names.Kind,
+			"apiVersion": crd.Spec.Group + "/v1",
+			"metadata": map[string]interface{}{
+				"name": name2,
+			},
+			"hostPort": "localhost:8080",
+		},
+	}
+
+	// After changing a CRD, the resources for versions will be re-created that can be result in
+	// cancelled connection (e.g. "grpc connection closed" or "context canceled").
```

Comment:
this shouldn't be true after https://github.com/kubernetes/kubernetes/pull/77816

Reply:
will leave it until #77816 is merged. Otherwise this flakes.

Reply:
It's already merged

Reply:
It merged. Removed this workaround.